### PR TITLE
Create, delete, and edit annotations

### PR DESCRIPTION
**Pull request by rebeccacremona**
This PR adds the backend routes for annotating a resource, as used by the Vue frontend.

It is not fully interchangeable with the Rails implementation: the Rails application calculates paragraph-based offsets for all annotations for historical reasons, and we have not ported that code to python. Paragraph-based offsets are still used in production during export. So, python-driven export has to be live before this code is made live. If we make this code live and then decide to switch back to the ruby... we'll have to come up with a plan for back-filling paragraph-based offsets annotations lacking any (which a) shouldn't happen, and b) shouldn't be hard, if it does indeed happen.)

For that to work.... I had to toggle some model fields to fields blank/null=True that are currently blank/null=False: `start_offset`, `end_offset`, and `start_paragraph`. For whatever reason, `end_paragraph` is already optional. This migration is NOT a no-op; it will actually change the DB. We'll have to decide when we want to run it..... As far as I can reason, it will be harmless to run, even while we are still using the Rails app.

Along for the ride: refactor the JS to get the Django CSRF token from a cookie instead of from the DOM.

### CODE REVIEW

**Comment by rebeccacremona** [comment]
I also commented out an unnecessary line in the ruby..... https://github.com/harvard-lil/h2o/pull/932/files#diff-1c6f0f7d733834ef819ef305d345a7e0L16

That might not be called for. It should speed up annotating, while we are still using the Rails app, but, since I guess there is a chance that I'm wrong and that seemingly-unused, unnecessary code is actually necessary... we could just leave it.

File: _python/main/views.py [hunk 1]
```diff
@@ -255,25 +257,118 @@ def render_with_actions(request, template_name, context=None, content_type=None,
...
-    resource = get_object_or_404(Resource.objects.select_related('casebook'), pk=resource_id)
 
-    # check permissions
-    if not resource.viewable_by(request.user):
-        return login_required_response(request)
+class AnnotationListView(APIView):
+
+    @method_decorator(perms_test(
+        {'args': ['resource'], 'results': {200: ['resource.casebook.owner', 'other_user', 'admin_user', None]}},
+        {'args': ['full_casebook_with_draft.drafts.resources.first'], 'results': {200: ['full_casebook_with_draft.drafts.resources.first.casebook.owner', 'admin_user'], 403: ['other_user'], 'login': [None]}},
```
Comment (jcushman) [inline _python/main/views.py]: Nit: the user here could be full_casebook_with_draft.drafts.owner

File: _python/conftest.py [hunk 2]
```diff
@@ -388,6 +395,12 @@ def factory(*ords):
         return casebook, sections_by_ordinal
     return factory
 
+@pytest.fixture
+def first_annotated_resource(full_casebook):
+    """
+    Given a casebook, returns its first annotated resource and annotation.
+    """
+    next(([r, r.annotations.first()] for r in full_casebook.resources if r.annotations.count() !=0), None)
```
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply _python/conftest.py]: Whoops, I meant to delete this. (I was intending to write a helper function that indeed does what the docstring says, and then ended up taking a different approach.)  💯 re: the saner way to look this up.... iterator + next not called for!

File: _python/conftest.py [hunk 3]
```diff
@@ -252,6 +252,13 @@ def admin_user(user_factory):
     user.roles.add(role)
     return user
 
+@pytest.fixture
+def published_annotation(db):
+    return ContentAnnotationFactory(resource=ResourceFactory())
+
+@pytest.fixture
+def private_annotation(db):
+    return ContentAnnotationFactory(resource=ResourceFactory(casebook=PrivateCasebookFactory()))
```
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply _python/conftest.py]: Thanks. Fixed. I'm not facile with these factories/fixtures/conventions yet: would love to chat at some point.

File: _python/main/views.py (hunk 1 above)
Comment (rebeccacremona) [reply _python/main/views.py]: 👍

**Pull request merged**